#!/bin/bash

########### The purpose of this script is to run QTL tools by gene to adjust for local ancestry at the TSS. 
### Is this perfect? No, but it's expected to be an improvement over just running nominal pass in terms of 
### refining the "lead variant" for each eGene. For use on a system other than slurm controller, job
### submission parameters might need to be renamed. 

wd="/path/to/where/your/input/is/local_anc"
vcfdir="/path/to/where/your/VCF/is/lamatrix"
genotypes="_filtered.vcf.gz"
prefix="UVA_coronary_v37_June2021_"
cov_file="/path/to/where/your/covariates/are/cov_filename.txt"
outdir="/path/to/where/your/output/goes/local_anc"

module load qtltools

cd ${wd}

a="your-account-name"
p="standard"
t="15:00"
m="8g"

# Because each gene is run as a separate job, only submit one chromosome at a time to avoid exceeding
# max # of job submissions on our server. Bright side, runs extremely fast this way
chr="1"

        while read gene; do 

sbatch --account=$a --mem=$m -p $p -t $t --wrap="QTLtools cis --vcf ${vcfdir}/UVA_b37_chr${chr}${genotypes} --bed ${wd}/chr${chr}_infile/${gene}.bed.gz --cov ${wd}/chr${chr}_infile/${gene}_${cov_file} --permute 100000 --region ${chr} --out ${outdir}/chr${chr}/QTLtools_LA_noPCs_${gene}.txt --seed 29048347 --window 500000" 

# List of genes for each chromosome was generated by selecting all genes for each chromosome 
# from the final combined bed file
        done < /project/cphg-millerlab/chani/QTL_input/chr${chr}_genes_for_la_qtltools.lst
